### PR TITLE
LTI 1.3 login support

### PR DESCRIPTION
**Pull request by lf-**
🎉

Note that this PR is based on #153.

This uses my library https://github.com/lf-/lti13 to provide LTI1.3 login. You can test it out using the reference tool at https://lti-ri.imsglobal.org/platforms/1255/resource_links with the resource link "Root".

To set it up, go to `/master_admin` and add a platform with the following settings [**Note: client_id is wrong in the screenshot, see below for the correct value**]: 
image

* issuer: aaaaa
* client_id: abcde
* redirect URL: https://lti-ri.imsglobal.org/platforms/1255/authorizations/new
* JWKs URL: https://lti-ri.imsglobal.org/platforms/1255/platform_keys/1248.json

We should probably handle non-emails in ident in the "email students" function, but as-is, it will just put values that are not emails (`randomThing@@https://canvas.instructure.com`) in there.

I have attempted to get this working with Canvas but hit an issue where our authentication cookies are not allowed to be set from within an iframe context due to their properties. I would just change the cookie settings, but I presume there may be some security implication of doing so, relating to CSRF, which I have not researched. See https://community.canvaslms.com/t5/Admin-Group/LTI-1-3-Open-in-new-tab/td-p/400128 for the thread on the Canvas forum about trying to get it to open in a new tab to resolve this. I have no reason to believe it will have any issues in Canvas once we figure out how to do the new-tab thing.

This PR is ready for review along with its friend, #153.

### CODE REVIEW

**Comment by lf-** [comment]
That CI failure seems random as I have not changed anything about the dependencies that just failed.

**Comment by lf-** [comment]
The oopsie in question:

2020-08-30T23:50:12-07:00

A resource: https://web.dev/samesite-cookies-explained/. It appears that we would need to set SameSite: none in order to have LTI authentication work within an iframe, which may be a bad idea as it is an anti-CSRF feature, but I have no idea what our risk exposure would be if we were to do that. Also, if we do that, SameSite=none also requires the Secure attribute, completely breaking sessions for local development, which is not acceptable.

Canvas could avoid hosing us with this by initiating with a GET rather than a POST if they're iframe'ing LTI stuff, but they do not :(

Settings for a test Canvas:

* issuer: https://canvas.instructure.com
* client_id: was 10000000000001 (it's the number on the developer key)
* oidc redirect: http://canvas.docker/api/lti/authorize_redirect
* jwks: http://canvas.docker/api/lti/security/jwks

Canvas side:
2020-08-31T00:47:23-07:00
2020-08-31T00:47:43-07:00
2020-08-31T00:48:00-07:00

**Comment by lf-** [comment]
I have fixed this issue by just changing the cookie configuration. I still am not sure if it is significant security exposure, however, it *does* work.

To mitigate issues with local development breakage, I have modified our approot implementation to make it static since it always had a value anyway, and it now detects an https approot and enables the necessary cookie settings for doing LTI in an iframe, but only in that case (since they were breaking if the Secure attribute was set otherwise).

I have tested locally as follows:

```
# in Carnap checkout in a nix-shell
$ (cd Carnap-Server; APPROOT=https://localhost:3000 DATAROOT=../dataroot BOOKROOT=../Carnap-Book/ PORT=2000 cabal run -f dev Carnap-Server)
# run mitmproxy for a second to generate certs then add ~/.mitmproxy/mitmproxy-ca-cert.pem to your browser's root certificate store (you can delete it fro the store after the test)
$ /usr/bin/mitmdump --mode reverse:http://localhost:2000/ -p 3000
```

Then configure Canvas or your other tools to use https://localhost:3000 instead of http://localhost:3000.

Result: (!!!)

2020-08-31T01:36:50-07:00

----

I think skipping the enrolment phase would be cool, and the information is all available in the token (see lti13 docs) to do so, but it will probably require some more schema and UI to make a mapping between the course info in the tokens to courses on a Carnap installation, which is beyond the scope of this PR. Also, we have public JWKs so we can poke the platform if we want to, including doing gradebook stuff, but that's going to require some spec reading and possibly some lti13 features.

Summary: remaining issue resolved. Good to go. I'd also suggest reviewing https://github.com/lf-/lti13 since it is effectively now Carnap code (has no other users).

**Comment by lf-** [comment]
Just did a rebase onto master. The diff should be a lot more legible now.

**Comment by lf-** [comment]
Another rebase to squash this commit and get it back up to speed with master.

**Comment by gleachkr** [comment]
Just built locally. Looks like oidc-client-0.5.1.0 might be needed as an extra-dep for stack.

File: stack.yaml [hunk 1]
```diff
@@ -18,7 +18,11 @@ extra-deps:
           - persistent-postgresql
           - persistent-template
           - persistent-qq
-# :(
+    - lti13-0.1.1.0
+    - yesod-auth-lti13-0.1.1.0
+
```
Comment (gleachkr) [inline stack.yaml:24]: I think we need oidc-client-0.5.1.0 here.

File: Carnap-Server/Foundation.hs [hunk 2]
```diff
@@ -295,20 +366,20 @@ instance YesodAuth App where
...
+                        Just _ -> setMessage "Now logged in"
 
     -- appDevel is a custom method added to the settings, which is true
     -- when yesod is running in the development environment and false
     -- otherwise
     authPlugins app = let settings = appSettings app in
                           if appDevel settings
-                              then [ authDummy ]
-                              else [ oauth2GoogleScoped ["email","profile"] (appKey settings) (appSecret settings) ]
+                              then [ authDummy, authLTI13 ]
```
Comment (gleachkr) [inline Carnap-Server/Foundation.hs]: It looks like lti13 might need a link for the AuthPlugin login? 

https://github.com/lf-/lti13/blob/c0aad6339c767d6c8e1780288da609915a87a89d/yesod-auth-lti13/src/Yesod/Auth/LTI13.hs#L325

At the moment, I'm not sure how to click through to attempt authentication there.

**Review by gleachkr — COMMENTED** [review]
Just a start, will pick up again later.

File: Carnap-Server/Foundation.hs (hunk 2 above)
Comment (lf-) [reply Carnap-Server/Foundation.hs]: This is actually bad copy on my part, not a technical error, but if you hit it, someone else definitely will too. It's just a note that you can go to your learning management system to initiate login. If you need a sample learning management system, there's the reference implementation, which there are details for in the first post (which you can use directly yourself; there's no login on it or anything). You can add the configuration details on the admin page.

File: stack.yaml (hunk 1 above)
Comment (lf-) [reply stack.yaml:24]: Very likely. Sorry about that, my stack doesn't work properly so I can't test full builds (that locale bug is still hitting me extremely badly, making it effectively not usable).

File: Carnap-Server/Foundation.hs (hunk 2 above)
Comment (lf-) [reply Carnap-Server/Foundation.hs]: As discussed on Matrix, this is resolved with both the addition of `authLTI13WithWidget` as of yesod-auth-lti13-0.1.2.0 and my attempt to improve the copy of the default.

Comment (gleachkr) [reply Carnap-Server/Foundation.hs]: Should we perhaps upgrade the deps on this PR, and use `authLTI13WithWidget`, rather than `authLTI13` in production?

Comment (lf-) [reply Carnap-Server/Foundation.hs]: Yep, we should come up with a nice sample screenshot and widget to show the user what we mean here. Maybe a "for more information" page? I'm not sure how to best do this. But we have the functionality to do it now :)

**Pull request merged**